moveit_core/robot_model/src/joint_model_group.cpp

```diff
@@ -97,188 +98,259 @@ bool jointPrecedes(const JointModel* a, const JointModel* b)
-      {
-        is_contiguous_index_list_ = false;
-        break;
-      }
-
-  // when updating/sampling a group state only, only mimic joints that have their parent within the group get updated.
-  for (const JointModel* mimic_joint : mimic_joints_)
-    // if the joint we mimic is also in this group, we will need to do updates when sampling
-    if (hasJointModel(mimic_joint->getMimic()->getName()))
-    {
-      int src = joint_variables_index_map_[mimic_joint->getMimic()->getName()];
-      int dest = joint_variables_index_map_[mimic_joint->getName()];
-      GroupMimicUpdate mu(src, dest, mimic_joint->getMimicFactor(), mimic_joint->getMimicOffset());
-      group_mimic_update_.push_back(mu);
-    }
+  constexpr auto is_contiguous = [](const auto& sub_view) {
+    const auto& values = sub_view | ranges::to<std::vector>();
+    return values[0] + 1 == values[1];
+  };
+  is_contiguous_index_list_ = !variable_index_list_.empty()
+    && ranges::all_of(variable_index_list_ | views::sliding(2), is_contiguous);
+
+  // when updating/sampling a group state only,
+  // only mimic joints that have their parent within the group get updated.
+  group_mimic_update_ = mimic_joints_
+    | views::filter([this](const auto& jm) { return hasJointModel(jm->getMimic()->getName()); })
+    | views::transform([this](const auto& jm) {
+      const int src = joint_variables_index_map_[jm->getMimic()->getName()];
+      const int dest = joint_variables_index_map_[jm->getName()];
+      return GroupMimicUpdate{src, dest, jm->getMimicFactor(), jm->getMimicOffset()};
+    })
+    | ranges::to<std::vector>();
 
   // now we need to make another pass for group links (we include the fixed joints here)
-  std::set<const LinkModel*> group_links_set;
-  for (const JointModel* joint_model : joint_model_vector_)
-    group_links_set.insert(joint_model->getChildLinkModel());
-  for (const LinkModel* group_link : group_links_set)
-    link_model_vector_.push_back(group_link);
-  std::sort(link_model_vector_.begin(), link_model_vector_.end(), OrderLinksByIndex());
-
-  for (const LinkModel* link_model : link_model_vector_)
-  {
-    link_model_map_[link_model->getName()] = link_model;
-    link_model_name_vector_.push_back(link_model->getName());
-    if (!link_model->getShapes().empty())
-    {
-      link_model_with_geometry_vector_.push_back(link_model);
-      link_model_with_geometry_name_vector_.push_back(link_model->getName());
-    }
-  }
-
-  // compute the common root of this group
-  if (!joint_roots_.empty())
-  {
-    common_root_ = joint_roots_[0];
-    for (std::size_t i = 1; i < joint_roots_.size(); ++i)
-      common_root_ = parent_model->getCommonRoot(joint_roots_[i], common_root_);
-  }
+  link_model_vector_ = joint_model_vector_
+    | views::transform([](const auto& jm) { return jm->getChildLinkModel(); })
+    | views::unique
+    | ranges::to<std::vector>()
+    | actions::sort(OrderLinksByIndex{});
+  link_model_map_ = link_model_vector_
```

Comment:
I wonder if it would improve readability if there were small comments on top of these operations. 
Here it could say `// Create a link name / link model map.`

Reply:
I added some comments where I thought the ranges expression was particularly complex.  Are there any others you think could be improved with comments?